src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1528 @@
+  // by const_iterator::operator==(const_iterator) plus implicit
+  // conversion of iterator to const_iterator.  But we need an
+  // additional overload to handle iterator == const_iterator when those
+  // types are different.
+
+  template<typename From, ENABLE_IF(is_convertible_iterator<From>())>
+  friend bool operator==(const From& lhs, const IteratorImpl& rhs) {
+    return rhs == lhs;
+  }
+
+  template<typename From, ENABLE_IF(is_convertible_iterator<From>())>
+  friend bool operator!=(const From& lhs, const IteratorImpl& rhs) {
+    return rhs != lhs;
+  }
+
+private:
+  // An iterator refers to either an object in the list, the root
+  // entry of the list, or null if singular.  See IteratorOperations
+  // for details of the encoding.
+  const void* _encoded_value;
+
+  // Allow explicit construction from an encoded const void*
+  // value.  But require exactly that type, disallowing any implicit
+  // conversions.  Without that restriction, certain kinds of usage
+  // errors become both more likely and harder to diagnose the
+  // resulting compilation errors.  [The remaining diagnostic
+  // difficulties could be eliminated by making EncodedValue a non-public
+  // class for carrying the encoded void* to iterator construction.]
+  template<typename EncodedValue,
+           ENABLE_IF(std::is_same<EncodedValue, const void*>::value)>
+  explicit IteratorImpl(EncodedValue encoded_value)
+    : _encoded_value(encoded_value)
+  {}
+};
+
+template<typename T,
+         IntrusiveListEntryAccessor<T> get_entry,
+         bool has_size>
+class IntrusiveList : public IntrusiveListImpl::SizeBase<has_size> {
+  // Give access to other instantiations, for splice().
+  template<typename U, IntrusiveListEntryAccessor<U>, bool>
+  friend class IntrusiveList;
+
+  // Give access for unit testing.
+  friend struct IntrusiveListImpl::TestSupport;
+
+  using Entry = IntrusiveListEntry;
+  using Impl = IntrusiveListImpl;
+  using ListTraits = Impl::ListTraits<T>;
+  using Super = Impl::SizeBase<has_size>;
+
+  // A subsequence of one list can be transferred to another list via splice
+  // if the lists have the same (ignoring const qualifiers) element type, use
+  // the same entry member, and either the receiver is a const-element list
+  // or neither is a const-element list.  A const element of a list cannot be
+  // transferred to a list with non-const elements.  That would effectively be
+  // a quiet casting away of const.  Assuming Other is a List, these
+  // constraints are equivalent to the constraints on conversion of
+  // Other::iterator -> iterator.  The presence or absence of constant-time
+  // size support for either of the lists doesn't affect whether splicing is
+  // permitted.
+  template<typename Other>
+  static constexpr bool can_splice_from() {
+    return Conjunction<Impl::IsListType<Other>,
+                       std::is_convertible<typename Other::iterator, iterator>>::value;
```

Comment:
Here and in can_swap we're using `Conjunction` rather than simple `&&`
expressions of the values. The intent is to delay the iterator check until
we've verified `Other` is actually a List type. But that isn't working.
Instead, if `Other` isn't a List then we get a syntax error on
`Other::iterator`, which is not at all helpful as error reporting goes. To fix
this we need to add a level of indirection so the code containing
`Other::iterator` doesn't get instantiated until the IsListType check has
passed.

Reply:
Fixed.